core/vm.cpp

```diff
@@ -1440,6 +1450,129 @@ class Interpreter {
         return decodeUTF8();
     }
 
+    const AST *regexMatch(const std::string &pattern, const std::string &string, bool full)
+    {
+        RE2 re(pattern, RE2::CannedOptions::Quiet);
+        if (!re.ok()) {
+            std::stringstream ss;
+            ss << "Invalid regex '" << re.pattern() << "': " << re.error();
+            throw makeError(stack.top().location, ss.str());
+        }
+
+        int num_groups = re.NumberOfCapturingGroups();
+
+        std::vector<std::string> rcaptures(num_groups);
```

Comment:
If the underlying library returns an empty string then it's clearly "OK" to do that.  Would we choose to differ?  I'm not sure there's much point, because you have to do `!= null` instead of `!= ''` right?  So it doesn't help the programmer to use `null` instead.  On the other hand, it means there is a difference between the underlying library and the Jsonnet manifestation of it.  That difference would have to be documented by us and remembered by everyone else.

Reply:
You could also make the argument that the only reason the RE2 C++ API returns the empty string is because it is much harder to return a sentinel in that language.  I wonder what the Go library does? @rohitjangid

Reply:
Well, empty match is not the same as the optional group is not matched.
Howewer, looks like both C++ re2 and golang regex implementations don't care, and leave default type value for such matches, unlike Rust (https://docs.rs/regex/latest/regex/struct.Captures.html#method.get) and js regexps.